lgc/patch/VertexFetch.cpp

```diff
@@ -475,6 +506,438 @@ bool LowerVertexFetch::runImpl(Module &module, PipelineState *pipelineState) {
+
+  // The first DWord
+  currentLoc = builder.CreateBitCast(currentLoc, PointerType::get(builder.getInt32Ty(), ADDR_SPACE_BUFFER_FAT_POINTER));
+  auto attr = builder.CreateLoad(builder.getInt32Ty(), currentLoc);
+
+  // The second DWord
+  currentLoc = builder.CreateGEP(builder.getInt32Ty(), currentLoc, {builder.getInt32(1)});
+  auto byteOffset = builder.CreateLoad(builder.getInt32Ty(), currentLoc);
+
+  // The third DWord
+  currentLoc = builder.CreateGEP(builder.getInt32Ty(), currentLoc, {builder.getInt32(1)});
+  auto instanceDivisor = builder.CreateLoad(builder.getInt32Ty(), currentLoc);
+
+  // The fourth DWord
+  currentLoc = builder.CreateGEP(builder.getInt32Ty(), currentLoc, {builder.getInt32(1)});
+  auto bufferFormat = builder.CreateLoad(builder.getInt32Ty(), currentLoc);
+
+  // attr[0~7]
+  auto descBinding = builder.CreateAnd(attr, builder.getInt32(0xFF));
+
+  // attr[8]
+  auto perInstance = builder.CreateAnd(attr, builder.getInt32(0x100));
+
+  // attr[10]
+  auto isPacked = builder.CreateAnd(attr, builder.getInt32(0x400));
+  isPacked = builder.CreateICmpNE(isPacked, zero);
+
+  // attr[12~15]
+  auto componentSize = builder.CreateIntrinsic(Intrinsic::amdgcn_ubfe, builder.getInt32Ty(),
+                                               {attr, builder.getInt32(12), builder.getInt32(4)});
+
+  auto xMask = builder.CreateAnd(attr, builder.getInt32(0x10000u));
+  auto yMask = builder.CreateAnd(attr, builder.getInt32(0x20000u));
+  auto zMask = builder.CreateAnd(attr, builder.getInt32(0x40000u));
+  auto wMask = builder.CreateAnd(attr, builder.getInt32(0x80000u));
+  xMask = builder.CreateICmpNE(xMask, zero);
+  yMask = builder.CreateICmpNE(yMask, zero);
+  zMask = builder.CreateICmpNE(zMask, zero);
+  wMask = builder.CreateICmpNE(wMask, zero);
+
+  // attr[20]
+  auto isBgr = builder.CreateAnd(attr, builder.getInt32(0x0100000));
+  isBgr = builder.CreateICmpNE(isBgr, zero);
+
+  // Load VbDesc
+  Value *vbDescPtr = builder.CreateGEP(vbDescTy, m_vertexBufTablePtr, descBinding);
+  LoadInst *loadInst = builder.CreateLoad(vbDescTy, vbDescPtr);
+  loadInst->setMetadata(LLVMContext::MD_invariant_load, MDNode::get(loadInst->getContext(), {}));
+  loadInst->setAlignment(Align(16));
+  Value *vbDesc = loadInst;
+  // Replace buffer format
+  vbDesc = builder.CreateInsertElement(vbDesc, bufferFormat, 3);
+
+  auto isPerInstance = builder.CreateICmpNE(perInstance, zero);
+
+  // PerInstance
+  auto vbIndexInstance = ShaderInputs::getInput(ShaderInput::InstanceId, builder, *m_lgcContext);
+  auto fInstanceDivisor = builder.CreateBitCast(instanceDivisor, builder.getFloatTy());
+  auto fInstance = builder.CreateSIToFP(vbIndexInstance, builder.getFloatTy());
+  fInstance = builder.CreateFMul(fInstance, fInstanceDivisor);
+  vbIndexInstance = builder.CreateFPToSI(fInstance, builder.getInt32Ty());
+  vbIndexInstance =
+      builder.CreateAdd(vbIndexInstance, ShaderInputs::getSpecialUserData(UserDataMapping::BaseInstance, builder));
```

Comment:
This will produce incorrect results for very large instance counts.

Reply:
Yes, you are right. This is an optimization of the division. 
We have not observed any incorrect cases so far, we have to change driver code If we want to fix this.

Reply:
> I've looked at the uber fetch shader commit in a bit more detail now and added comments.
> 
> Vulkan supports 10_10_10_2 formats for which some HW workarounds are required to get correct values in the W channel.

It seems to be an issue before gfx9.  Uber shader is only for pipeline library, for these corner cases, we ignore it for now.